core/src/main/java/io/questdb/griffin/SqlCompiler.java

```diff
@@ -557,6 +564,29 @@ public static RecordToRowCopier assembleRecordToRowCopier(BytecodeAssembler asm,
                             asm.invokeInterface(wPutByte, 2);
                             break;
                         case ColumnType.FLOAT:
+                            asm.dup2();
```

Comment:
I wonder if this byte code could be static function copier would invoke ?

Reply:
Hi Vlad, I guess it could be a static function call (per data-type), something like :

void checkBounds( double value, double min, double max, int position) {
  if ( value < min || value > max ){
     throw new SqlException.inconvertibleValue("Value " + value + " at position " + position + " is not between " + min + " and " + max );
  }
}
etc. 
at the cost of function call . I don't know if that's acceptable considering  . 
It would still be good to enrich exception with column name at a higher level ( unless there's a way to extract metadata from copy method arguments .

Reply:
@bluestreak01 Should I replace bytecode with static method call and add validation to all narrowing primitive type conversions within  assembleRecordToRowCopier ? 
Should I also validate long -> float/double conversion ?

Reply:
hi Bolek, yes, I would suggest to replace bytecode with static method calls where possible. JIT will inline them anyway without us trying to be very granular with the bytecode.

long->float conversion  should be working I suppose, does it not?

Reply:
In web console: 
max long to double : 
select cast(9223372036854775807 as double) ;
9223372036854776000

max long to float : 
select cast(9223372036854775807 as float) ;
9223372000000000000

Reply:
this looks correct to me

Reply:
What I meant is that such conversion loses precision because such long can't be represented as Double so it's not exactly a safe widening conversion (but that's how things work) . 
Anyway, I applied changes you requested and now there are failing tests : 

  SqlCompilerTest.testCastIntByte » Sql [0] inconvertible value: -2147483648 [IN...
  SqlCompilerTest.testCastLongShort » Sql [0] inconvertible value: NaN [LONG -> ...
  SqlCompilerTest.testCastIntShort » Sql [0] inconvertible value: -2147483648 [I...
  SqlCompilerTest.testCastLongByte » Sql [0] inconvertible value: NaN [LONG -> B...
  SqlCompilerTest.testCastLongInt » Sql [0] inconvertible value: NaN [LONG -> IN...
  SqlCompilerTest.testCastShortByte » Sql [0] inconvertible value: 1430 [SHORT -...

They fail for 2 reasons : 
1. input value is outside of allowed range  for target type
2. Long.MIN_VALUE and Integer.MIN_VALUE are interpreted as NaN in qdb code . Should these NaNs be processed in a different way ?  It seems there's no NaN in smaller integer types .

Additional question -> should Float/Double NaN translate to Long/Int NaN in implicit conversion ?

PS: It would make sense to document these Long/Int NaNs on https://questdb.io/docs/reference/sql/datatypes/ because e.g. NaN is returned as null in web console .

Reply:
Hi Bolek,

> Long.MIN_VALUE and Integer.MIN_VALUE are interpreted as NaN in qdb code . Should these NaNs be processed in a different way ? It seems there's no NaN in smaller integer types .

Yes, `int` and `long` NaN values always have special treatment in questdb codebase

> Additional question -> should Float/Double NaN translate to Long/Int NaN in implicit conversion ?

Absolutely

> You might consider allowing narrowing conversions going in the other direction (with over/under-flow check ) to make things more symmetric .

Overflow check is computationally expensive and is not necessary for explicit conversions. Type narrowing is also often used to get lower bits of the greater number. It is possibly to introduce a group functions (other than cast) that would produce NULL in case of overflow.

> PS: It would make sense to document these Long/Int NaNs on https://questdb.io/docs/reference/sql/datatypes/ because e.g. NaN is returned as null in web console .

Yes, it would. Tagging @bsmth here to make sure this is on our radar

Reply:
Hi Vlad,

>>Additional question -> should Float/Double NaN translate to Long/Int NaN in implicit conversion ?
>Absolutely

Currently  intrinsic conversions ( f2i, f2l, d2i, d2l ) translate NaN to 0  . 
Should I add something like : 
value == value ? (int) value : Numbers.INT_NaN
to relevant conversions in SqlCompiler.assembleRecordToRowCopier() in this PR or should it be a separate issue ?
If so - as bytecode (would require stackmap) or within addCheckIntBoundsCall (could return adjusted value and type-specific functions would be required) ?
  
From what I've seen explicit conversions are dealt with before row data is passed to copier so they won't trigger range checks - at least for : 
create table tabi( i int); 
insert into tabi values ( cast ( 123.1 as int ) );
and 
create table tabi( i int); 
create table tabd (d double);
insert into tabd( 123.1 );
insert into tabi select cast ( d as int ) from tabd ;

Reply:
Hi Vlad, 
Last questions (I hope).
1. Exception message produced by bounds check isn't too helpful at the moment so I'd like to add : 
sql position(optional) 
from column name
to column name
Looking at RecordToRowCopier interface I don't see a straightforward way to extract the data from method input parameters . 
Can I add the following fields to rowcopier : 
InsertModel (optional)
input record metadata 
writer metadata
and use in checkDoubleBounds/checkLongBounds ? 

2. Are current addCheckXXXBoundsCall methods ok with widening conversions (f2d, i2l) or should I remove conversions and create method for each primitive type ?